### PR TITLE
add key points to episode 3

### PR DESCRIPTION
**Pull request by raynamharris**
Two of the key points I took almost directly from the text. For the third, I tried to think of a bigger picture key point. Maybe these are useful.

### CODE REVIEW

**Comment by wrightaprilm** [comment]
Thanks for this! Happy to merge if it looks good to you, @maxim-belkin.

File: _episodes/03-index-slice-subset.md [hunk 1]
```diff
@@ -16,7 +16,10 @@ objectives:
     - "Locate subsets of data using masks."
     - "Describe BOOLEAN objects in Python and manipulate data using BOOLEANs."
 keypoints:
-    - "FIXME"
+    - "Data can be subset using column headings, numeric ranges , operators, or specific index locations."
```
Comment (maxim-belkin) [inline _episodes/03-index-slice-subset.md]: No space before comma in `numeric ranges , `

File: _episodes/03-index-slice-subset.md [hunk 2]
```diff
@@ -16,7 +16,10 @@ objectives:
...
+    - "Python uses 0-based indexing."
```
Comment (maxim-belkin) [inline _episodes/03-index-slice-subset.md]: I suggest either adding a brief explanation or replacing the keypoint with it.
```
Python uses 0-based indexing: <brief explanation>
```
Explanation could be:
```
The first element in a list, tuple or any other data structure has an index of 0
```

File: _episodes/03-index-slice-subset.md [hunk 3]
```diff
@@ -16,7 +16,10 @@ objectives:
...
+    - "Data indexing, slicing and subsetting are common steps in data wrangling and exploration."
```
Comment (maxim-belkin) [inline _episodes/03-index-slice-subset.md]: what message are you trying to convey with this keypoint? It seems too general and not Python-specific

File: _episodes/03-index-slice-subset.md (hunk 1 above)
Comment (maxim-belkin) [reply _episodes/03-index-slice-subset.md]: another point: it's not super clear what `Data can be subset` means.

**Comment by maxim-belkin** [comment]
I've submitted a few comments. Let me know if you do or do not see them
Thanks, @raynamharris!

File: _episodes/03-index-slice-subset.md (hunk 1 above)
Comment (wrightaprilm) [reply _episodes/03-index-slice-subset.md]:
```suggestion
    - "Data can be subset using column headings, numeric ranges, operators, or specific index locations."
```

File: _episodes/03-index-slice-subset.md (hunk 2 above)
Comment (wrightaprilm) [reply _episodes/03-index-slice-subset.md]:
```suggestion
    - "Python uses 0-based indexing, in which the first element in a list, tuple or any other data structure has an index of 0."
```

File: _episodes/03-index-slice-subset.md (hunk 3 above)
Comment (wrightaprilm) [reply _episodes/03-index-slice-subset.md]: I like that - this is the Python lesson, but tapping into a larger set of tasks and concerns.

**Comment by wrightaprilm** [comment]
And I added a couple suggestions. If they look good to you, Rayna, you can accept them and leave comment. Otherwise, feel free to suggest other wordings.

File: _episodes/03-index-slice-subset.md (hunk 1 above)
Comment (raynamharris) [reply _episodes/03-index-slice-subset.md]: Very good point. Now I can't remember why I wanted to add this particular keypoint. Most of the lesson is about using column names to summarize data. Because really, this keypoint summaries this [quick guide to subsetting data in R](https://www.statmethods.net/management/subset.html).

I'm happy to remove it if you think it's irrelevant.

Comment (raynamharris) [reply _episodes/03-index-slice-subset.md]: space deleted in https://github.com/datacarpentry/python-ecology-lesson/pull/322/commits/034f56aaad5f0332b02d7d7cd1d38238ba8d1fc0

File: _episodes/03-index-slice-subset.md (hunk 2 above)
Comment (raynamharris) [reply _episodes/03-index-slice-subset.md]: thanks april! suggestion incorporated in https://github.com/datacarpentry/python-ecology-lesson/pull/322/commits/034f56aaad5f0332b02d7d7cd1d38238ba8d1fc0

File: _episodes/03-index-slice-subset.md (hunk 3 above)
Comment (raynamharris) [reply _episodes/03-index-slice-subset.md]: Thanks April. Yes, I was specifically trying to make a broader picture of why this is important. Maybe is could say  something like "The `pandas` library useful for data indexing, slicing and subsetting, which are common steps in data wrangling and exploration."

Comment (maxim-belkin) [reply _episodes/03-index-slice-subset.md]: 1. I think keypoints should be specific to episodes
2. Indexing and slicing are not Pandas-specific.

Perhaps, you meant something like this
```
For data exploration and wrangling in Python you can use such powerful subsetting methods as indexing (accessing individual elements of complex data structures _via_ their index) and slicing (accessing several individual elements...)
```

File: _episodes/03-index-slice-subset.md (hunk 1 above)
Comment (wrightaprilm) [reply _episodes/03-index-slice-subset.md]: This seems to. have bled into the next point, so I'm going to mark this as resolved.

File: _episodes/03-index-slice-subset.md (hunk 3 above)
Comment (wrightaprilm) [reply _episodes/03-index-slice-subset.md]: I don't disagree, but the keypoint is specific to what we're doing in the lesson. If we want it to be more pythonic, I don't not like what you've written, but its a bit wordy. Perhaps

```suggestion
    - "Pandas enables common data exploration steps such as data indexing, slicing and subsetting."
```

File: _episodes/03-index-slice-subset.md (hunk 1 above)
Comment (maxim-belkin) [reply _episodes/03-index-slice-subset.md]: My point was that in this lesson we use column headings, ranges, etc for _accessing_ portions of data not for changing it. Because the verb `subset` means `be a subset of`, the phrase `Data can be subset` can be understood as `Data can be changed in-place`

File: _episodes/03-index-slice-subset.md (hunk 3 above)
Comment (maxim-belkin) [reply _episodes/03-index-slice-subset.md]: Python native lists and tuples can be indexed and sliced too.
Pandas and numpy can do condition-based subsetting.

File: _episodes/03-index-slice-subset.md (hunk 1 above)
Comment (wrightaprilm) [reply _episodes/03-index-slice-subset.md]: Oh, interesting, I had not considered this interpretation. Is there a wording that might be better?

File: _episodes/03-index-slice-subset.md (hunk 3 above)
Comment (wrightaprilm) [reply _episodes/03-index-slice-subset.md]: Sure, but I don't think it's on us to emphasize the uniqueness of the technology we teach.

Comment (maxim-belkin) [reply _episodes/03-index-slice-subset.md]: `Pandas supports...` would make this keypoint more accurate.

On a different note: aren't indexing and slicing examples of subsetting? Or did we miss `conditional` before `subsetting`?

Comment (maxim-belkin) [reply _episodes/03-index-slice-subset.md]: I think we ought to change line 37 to `condition-based` or `conditional` subsetting

File: _episodes/03-index-slice-subset.md (hunk 1 above)
Comment (maxim-belkin) [reply _episodes/03-index-slice-subset.md]: Nah :)


<br>
jk! I'm sure there is. Something like


>In Python, portions of data can be accessed using indices, slices, column headings (Panda's data frames), and condition-based subsetting (Panda's data frames and numpy arrays).

File: _episodes/03-index-slice-subset.md (hunk 3 above)
Comment (raynamharris) [reply _episodes/03-index-slice-subset.md]: yep! i love this new beta feature

Comment (wrightaprilm) [reply _episodes/03-index-slice-subset.md]: It really is great. It's exactly what I need for working with people who are new to LaTeX, too. Leaving line edits is always the biggest stumbling block for the new folks.

File: _episodes/03-index-slice-subset.md (hunk 1 above)
Comment (raynamharris) [reply _episodes/03-index-slice-subset.md]: LGTM

Comment (raynamharris) [reply _episodes/03-index-slice-subset.md]: This beta suggest and apply feature is awesome!

**Comment by wrightaprilm** [comment]
I think we're all resolved. If you're happy, I'm happy, @maxim-belkin.

File: _episodes/03-index-slice-subset.md (hunk 1 above)
Comment (maxim-belkin) [reply _episodes/03-index-slice-subset.md]: > Does this suggestion work for everyone?

Looks good as long as all of us understand what `condition-based subsetting` is :)

**Comment by maxim-belkin** [comment]
Besides (really minor) https://github.com/datacarpentry/python-ecology-lesson/pull/322#discussion_r231645115, this PR looks great! Great job both of you! https://youtu.be/69sa-jYk444?t=2

File: _episodes/03-index-slice-subset.md (hunk 1 above)
Comment (wrightaprilm) [reply _episodes/03-index-slice-subset.md]:
```suggestion
    - "In Python, portions of data can be accessed using indices, slices, column headings, and condition-based subsetting."
```

I'll add it in - the beta suggestions thing is a little finnicky to if you changed a prior line.

File: _episodes/03-index-slice-subset.md [hunk 4]
```diff
@@ -16,7 +16,10 @@ objectives:
     - "Locate subsets of data using masks."
     - "Describe BOOLEAN objects in Python and manipulate data using BOOLEANs."
 keypoints:
-    - "FIXME"
+    - "Data can be subset using column headings, numeric ranges, operators, or specific index locations."
```
Comment (wrightaprilm) [inline _episodes/03-index-slice-subset.md]:
```suggestion
     - "In Python, portions of data can be accessed using indices, slices, column headings, and condition-based subsetting."
```

**Comment by wrightaprilm** [comment]
Ah, yes, the comment - I didn't realize that my suggestion became inapplicable when a previous line was edited. I made a new suggestion containing the language we agreed on (with the condition-based data selection). Let me know if its not visible, @raynamharris. Once that's sewn up, I'll merge.

**Comment by raynamharris** [comment]
No worries. Thanks to both of you for all the discussion. I hope these updates are useful for learners and trainers :)
